packages/core/src/keys.ts

```diff
@@ -4,35 +4,8 @@
 // License text available at https://opensource.org/licenses/MIT
 
 export namespace CoreBindings {
-  // application-wide bindings
-
-  export const HTTP_PORT = 'http.port';
-  export const HTTP_HANDLER = 'http.handler';
-
-  export const API_SPEC = 'application.apiSpec';
-  export const SEQUENCE = 'sequence';
-
-  export namespace SequenceActions {
-    export const FIND_ROUTE = 'sequence.actions.findRoute';
-    export const PARSE_PARAMS = 'sequence.actions.parseParams';
-    export const INVOKE_METHOD = 'sequence.actions.invokeMethod';
-    export const LOG_ERROR = 'sequence.actions.logError';
-    export const SEND = 'sequence.actions.send';
-    export const REJECT = 'sequence.actions.reject';
-  }
-
-  export const GET_FROM_CONTEXT = 'getFromContext';
-  export const BIND_ELEMENT = 'bindElement';
```

Comment:
The current implementation of these providers, in `/rest/router/providers` are tightly-coupled to HTTP:

```ts
export class GetFromContextProvider implements Provider<GetFromContext> {
  constructor(@inject(RestBindings.Http.CONTEXT) protected context: Context) {}

  value() {
    return (key: string) => this.context.get(key);
  }
}
```

Admittedly, I'm still a little fuzzy as to where this is being leveraged _exactly_, but if I see a way to move it back to core, I'll do it (or if you know exactly what to change, I'll do that instead).

Reply:
Ah, this is an important topic to discuss. The problem basically boils down to whether it's a good idea to allow user code to receive the current context via dependency injection. Originally, I was reluctant to allow this, because then we are making it too easy to use the context as a service locator by calling `ctx.get` directly, instead of pushing people to use proper dependency injection and let the framework to resolve the dependencies from the context.

As a middle ground, `getFromContext` and `bindElement` were implemented at request-level scope only.

I see two options how to make them app-wide (available for any context):

 - Move their implementation to `@loopback/context`, bind these two values in `Context` constructor. Then in theory we should not need to expose current context at all. (But there may be other places e.g. in the router where we may still legitimately need to get the current context injected, IDK.)

    ```ts
    class Context {
      constructor(/*...*/) {
        // ...
        this.bind(ContextBindings.getFromContext)
          .toDynamicValue((key: string) => this.get(key));
      }
    }
    ```

    If we go down this route, then I think we should rename `bindElement` to `bindToContext` for consistency with `getFromContext`, and also because there is no notion of actions and elements within the context module, actions and elements are higher-level concepts.

 - Give up the quest for purity and accept the fact that there are legitimate situations where the code requires the full current context. Add a core/context binding allowing retrieval of the current context via injection

    ```ts
    class Context {
      constructor(/*...*/) {
        // ...
        this.bind(ContextBindings.CURRENT_CONTEXT).to(this);
      }
    }
    ```

    Then we can bind `getFromContext` and `bindElement` at application level, because their providers can receive the current context via `ContextBindings.CURRENT_CONTEXT` (similarly to how they are receiving `RestBindings.Http.CONTEXT` now).

BTW, I think the `Http` namespace is no longer necessary, we can list all constants directly in `RestBinding` root namespace (e.g. `RestBindings.CONTEXT`).

@raymondfeng @virkt25 thoughts?

Reply:
> I personally like the simpler approach of using `ctx.get` over `@inject`. If this is made available app wide are there any downsides for the user using the application? (performance? other issues?). I can't think of any and hence an in favour of making this app level. 

My main issue with service locators is that classes obtaining dependencies from a service locator make it difficult to see what dependencies they require and the compiler cannot verify that all dependencies were satisfied. As a result, pure unit-tests are difficult to write and maintain.

Compare these two examples:

Service-locator:

```ts
class ProductController {
  constructor(protected ctx: Context) {
  }

  async list() {
    const repo = await ctx.get('repositories.Product');
    return await repo.find();
  }

  async create() {
    const logger = await ctx.get('logger');
    logger.error('Product.create is not implemented yet.');
    throw new HttpError.NotImplemented();
  }
}

it('logs an error on create', async () => {
  const ctx = new Context();
  // setup dependencies - but what are they?
  const controller = new ProductController(ctx);
  // etc.
});
```

Dependency-injection via constructor arguments:

```ts
class ProductController {
  constructor(protected repo: ProductRepository, protected logger: Logger) {
  }

  async list() {
    return await this.repo.find();
  }

  async create() {
    this.logger.error('Product.create is not implemented yet.');
    throw new HttpError.NotImplemented();
  }
}

it('logs an error on create', async () => {
  const logger = // create logger stub
  const repo = // create repo stub
  const controller = new ProductController(repo, logger);
  // etc.
});
```

Now imagine you are adding a third dependency to this controller class. With dependency-injection, the compiler will show you all places where you need to add a new dependency. When using service-locator, the controller API remains the same, therefore you will discover the missing dependency only after you have run your tests, and only in the places covered by your tests.

Reply:
On the second thought, and considering that we already have `inject.getter()` and `inject.setter()` to support classes that need defer resolution of a binding or (re)bind a value, I think we should go with the simplest solution, which is `ContextBindings.CURRENT_CONTEXT`, and consider to remove `getFromContext` and `bindElement` completely.